File: backend/ocean_supply.py
from __future__ import annotations

import json
import math
import statistics
import time
import urllib.error
import urllib.request
from dataclasses import dataclass, asdict
from pathlib import Path
from typing import Any, Dict, Iterable, List, Optional, Tuple

DEFAULT_DASHBOARD_ENDPOINTS = [
    "https://api.oncompute.ai/envs",
    "https://api.oncompute.ai/nodes",
    "https://api.oncompute.ai/summary",
    "https://api.oncompute.ai/all-summary",
    "https://analytics.oncompute.ai/global-stats",
    "https://analytics.oncompute.ai/gpu-popularity",
]

ROOT = Path(__file__).resolve().parents[1]
SAMPLE_PATH = ROOT / "data" / "sample_supply.json"
NODE_ENDPOINTS_PATH = ROOT / "data" / "node_endpoints.txt"


@dataclass
class SourceResult:
    name: str
    state: str
    message: str
    url: Optional[str] = None


@dataclass
class ComputeResource:
    timestamp: str
    source: str
    providerId: str
    providerLabel: str
    nodeEndpoint: str
    environmentId: str
    region: str
    resourceType: str
    resourceName: str
    total: float
    inUse: float
    available: float
    feeToken: str
    pricePerMinute: Optional[float]
    pricePerHour: Optional[float]
    status: str
    raw: Dict[str, Any]


def utc_now() -> str:
    return time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())


def http_get_json(url: str, timeout: float = 8.0) -> Any:
    request = urllib.request.Request(url, headers={"User-Agent": "FishDashboard/0.1"})
    with urllib.request.urlopen(request, timeout=timeout) as response:
        text = response.read().decode("utf-8", errors="replace")
    return json.loads(text)


def load_sample() -> Dict[str, Any]:
    return json.loads(SAMPLE_PATH.read_text())


def read_node_endpoints() -> List[str]:
    if not NODE_ENDPOINTS_PATH.exists():
        return []
    endpoints = []
    for line in NODE_ENDPOINTS_PATH.read_text().splitlines():
        line = line.strip()
        if not line or line.startswith("#"):
            continue
        endpoints.append(line.rstrip("/"))
    return endpoints


def walk_json(obj: Any) -> Iterable[Dict[str, Any]]:
    """Yield dict nodes from an arbitrary JSON shape."""
    if isinstance(obj, dict):
        yield obj
        for value in obj.values():
            yield from walk_json(value)
    elif isinstance(obj, list):
        for item in obj:
            yield from walk_json(item)


def pick_first(d: Dict[str, Any], keys: List[str], default: Any = None) -> Any:
    for key in keys:
        if key in d and d[key] not in (None, ""):
            return d[key]
    return default


def to_float(value: Any, default: float = 0.0) -> float:
    try:
        if value is None:
            return default
        if isinstance(value, str):
            cleaned = value.replace("$", "").replace("/hr", "").replace(",", "").strip()
            return float(cleaned)
        return float(value)
    except Exception:
        return default


def looks_like_gpu_name(value: str) -> bool:
    value = value.lower()
    return any(token in value for token in ["h200", "h100", "a100", "a6000", "l40", "gpu", "rtx", "mi300"])


def infer_resource_type(row: Dict[str, Any]) -> str:
    explicit = str(pick_first(row, ["resourceType", "type", "kind", "resource_type"], "")).lower()
    if "gpu" in explicit:
        return "gpu"
    if "cpu" in explicit:
        return "cpu"
    if "ram" in explicit or "memory" in explicit:
        return "ram"
    if "disk" in explicit or "storage" in explicit:
        return "disk"
    name = str(pick_first(row, ["resource", "resourceName", "name", "description", "gpu", "gpuType", "gpu_type"], ""))
    if looks_like_gpu_name(name):
        return "gpu"
    return "unknown"


def normalize_price(row: Dict[str, Any]) -> Tuple[Optional[float], Optional[float]]:
    # Prefer explicit hourly fields if present.
    hourly = pick_first(row, ["pricePerHour", "price_per_hour", "usdHr", "usd_hr", "priceUsdHr", "price"], None)
    if hourly is not None:
        value = to_float(hourly, math.nan)
        if not math.isnan(value):
            # If very tiny value, it might be per minute. Otherwise treat as hourly.
            if value < 0.5:
                return value, value * 60
            return value / 60, value

    minute = pick_first(row, ["pricePerMinute", "price_per_minute", "rate", "minutePrice"], None)
    if minute is not None:
        value = to_float(minute, math.nan)
        if not math.isnan(value):
            return value, value * 60

    fees = pick_first(row, ["fees", "pricing", "prices"], None)
    if isinstance(fees, list):
        for fee in fees:
            if isinstance(fee, dict):
                minute = pick_first(fee, ["price", "pricePerMinute", "amount"], None)
                if minute is not None:
                    value = to_float(minute, math.nan)
                    if not math.isnan(value):
                        # Ocean fees are often per minute in configured resource units.
                        return value, value * 60
    elif isinstance(fees, dict):
        minute = pick_first(fees, ["price", "pricePerMinute", "amount"], None)
        if minute is not None:
            value = to_float(minute, math.nan)
            if not math.isnan(value):
                return value, value * 60

    return None, None


def normalize_dicts(payload: Any, source_name: str, source_url: str) -> List[ComputeResource]:
    timestamp = utc_now()
    resources: List[ComputeResource] = []
    for d in walk_json(payload):
        resource_type = infer_resource_type(d)
        if resource_type == "unknown":
            continue

        resource_name = str(pick_first(d, ["resourceName", "resource", "name", "description", "gpu", "gpuType", "gpu_type", "type"], resource_type.upper()))
        total = to_float(pick_first(d, ["total", "count", "quantity", "max", "availableTotal", "gpus"], 0), 0)
        in_use = to_float(pick_first(d, ["inUse", "in_use", "used", "running", "busy"], 0), 0)
        available = to_float(pick_first(d, ["available", "free", "remaining"], max(total - in_use, 0)), max(total - in_use, 0))
        if total <= 0 and available > 0:
            total = available + in_use
        if total <= 0 and resource_type == "gpu":
            total = 1
            available = max(1 - in_use, 0)

        ppm, pph = normalize_price(d)
        provider_id = str(pick_first(d, ["providerId", "provider", "owner", "ownerWallet", "wallet", "nodeId", "id"], source_name))
        env_id = str(pick_first(d, ["environmentId", "envId", "id", "name"], "unknown-env"))
        region = str(pick_first(d, ["region", "location", "country", "zone"], "unknown"))
        fee_token = str(pick_first(d, ["feeToken", "token", "paymentToken", "symbol"], "unknown"))

        resources.append(
            ComputeResource(
                timestamp=timestamp,
                source=source_name,
                providerId=provider_id,
                providerLabel=provider_id[:18],
                nodeEndpoint=source_url,
                environmentId=env_id,
                region=region,
                resourceType=resource_type,
                resourceName=resource_name,
                total=total,
                inUse=in_use,
                available=available,
                feeToken=fee_token,
                pricePerMinute=ppm,
                pricePerHour=pph,
                status="available" if available > 0 else "busy" if total > 0 else "unknown",
                raw=d,
            )
        )
    return resources


def fetch_direct_node(endpoint: str) -> Tuple[List[ComputeResource], SourceResult]:
    url = f"{endpoint.rstrip('/')}/api/services/computeEnvironments"
    try:
        payload = http_get_json(url)
        resources = normalize_dicts(payload, "direct-node", endpoint)
        return resources, SourceResult(name="direct-node", state="live", message=f"Fetched {len(resources)} resources", url=url)
    except Exception as exc:
        return [], SourceResult(name="direct-node", state="unavailable", message=str(exc), url=url)


def fetch_dashboard_endpoint(url: str) -> Tuple[List[ComputeResource], SourceResult]:
    try:
        payload = http_get_json(url)
        resources = normalize_dicts(payload, "dashboard-api", url)
        return resources, SourceResult(name="dashboard-api", state="live", message=f"Fetched {len(resources)} resources", url=url)
    except Exception as exc:
        return [], SourceResult(name="dashboard-api", state="unavailable", message=str(exc), url=url)


def summarize(resources: List[ComputeResource], sources: List[SourceResult], state: str) -> Dict[str, Any]:
    gpu_resources = [r for r in resources if r.resourceType == "gpu"]
    provider_ids = {r.providerId for r in gpu_resources}
    total_gpus = sum(r.total for r in gpu_resources)
    available_gpus = sum(r.available for r in gpu_resources)
    h200_prices = [r.pricePerHour for r in gpu_resources if r.pricePerHour is not None and "h200" in r.resourceName.lower()]
    h200_from = min(h200_prices) if h200_prices else None

    grouped: Dict[str, List[ComputeResource]] = {}
    for r in gpu_resources:
        key = r.resourceName or "GPU"
        grouped.setdefault(key, []).append(r)

    gpu_supply = []
    for gpu, rows in grouped.items():
        prices = [r.pricePerHour for r in rows if r.pricePerHour is not None]
        gpu_supply.append({
            "gpu": gpu,
            "total": sum(r.total for r in rows),
            "available": sum(r.available for r in rows),
            "providers": len({r.providerId for r in rows}),
            "lowestUsdHr": min(prices) if prices else None,
            "medianUsdHr": statistics.median(prices) if prices else None,
            "regions": sorted({r.region for r in rows if r.region}),
        })

    providers = []
    provider_groups: Dict[str, List[ComputeResource]] = {}
    for r in gpu_resources:
        provider_groups.setdefault(r.providerId, []).append(r)
    for provider_id, rows in provider_groups.items():
        prices = [r.pricePerHour for r in rows if r.pricePerHour is not None]
        providers.append({
            "providerId": provider_id,
            "label": rows[0].providerLabel,
            "region": rows[0].region,
            "gpuTypes": sorted({r.resourceName for r in rows}),
            "availableGpus": sum(r.available for r in rows),
            "lowestUsdHr": min(prices) if prices else None,
            "uptime7d": None,
            "benchmarkStatus": "needs benchmark",
            "pilotEligible": sum(r.available for r in rows) > 0,
        })

    return {
        "dataState": state,
        "lastUpdated": utc_now(),
        "sources": [asdict(s) for s in sources],
        "kpis": {
            "totalGpus": total_gpus,
            "availableGpus": available_gpus,
            "providerCount": len(provider_ids),
            "h200FromUsdHr": h200_from,
            "oceanNativeJobs": 0,
            "providerPayoutUsd": 0,
        },
        "gpuSupply": gpu_supply,
        "providers": providers,
    }


def collect_summary() -> Dict[str, Any]:
    resources: List[ComputeResource] = []
    sources: List[SourceResult] = []

    for url in DEFAULT_DASHBOARD_ENDPOINTS:
        rows, source = fetch_dashboard_endpoint(url)
        sources.append(source)
        resources.extend(rows)

    for endpoint in read_node_endpoints():
        rows, source = fetch_direct_node(endpoint)
        sources.append(source)
        resources.extend(rows)

    if resources:
        return summarize(resources, sources, "live")

    sample = load_sample()
    sample["sources"] = [asdict(s) for s in sources] + sample.get("sources", [])
    sample["lastUpdated"] = utc_now()
    sample["dataState"] = "sample"
    return sample


def collect_resources() -> Dict[str, Any]:
    summary = collect_summary()
    # In sample mode, resources may not exist. Convert provider/gpu rows loosely.
    return {"dataState": summary.get("dataState", "sample"), "resources": []}


if __name__ == "__main__":
    print(json.dumps(collect_summary(), indent=2))

File: src/lib/oceanSupply.ts
import { readFile } from "node:fs/promises";
import path from "node:path";
import { collectProviderPilotRegistry, type ProviderPilotRegistry } from "@/lib/providerPilot";
import type {
  ComputeResource,
  DataState,
  GpuSupplyRow,
  OceanData,
  OceanSummary,
  ProviderScore,
  SourceResult
} from "@/lib/types";

const ROOT = process.cwd();
const SAMPLE_PATH = path.join(ROOT, "data", "sample_supply.json");
const NODE_ENDPOINTS_PATH = path.join(ROOT, "data", "node_endpoints.txt");
const REQUEST_TIMEOUT_MS = 8000;
const MAX_PAGES = Number(process.env.ONCOMPUTE_MAX_PAGES ?? "3");

const BASE_USDC = "0x833589fcD6EDb6E08f4c7C32D4f71b54bdA02913".toLowerCase();

type JsonRecord = Record<string, unknown>;

type NodeMetadata = {
  providerId: string;
  providerLabel: string;
  region: string;
  eligible: boolean | null;
  eligibilityCause: string | null;
  http: boolean | null;
  p2p: boolean | null;
  version: string | null;
  lastSeen: string | null;
  uptimeSeconds: number | null;
  nodeEndpoint: string;
};

const nowIso = () => new Date().toISOString();

function isRecord(value: unknown): value is JsonRecord {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function asArray(value: unknown): unknown[] {
  return Array.isArray(value) ? value : [];
}

function asString(value: unknown, fallback = ""): string {
  return typeof value === "string" && value.trim() ? value : fallback;
}

function asNumber(value: unknown, fallback = 0): number {
  if (typeof value === "number" && Number.isFinite(value)) {
    return value;
  }
  if (typeof value === "string" && value.trim()) {
    const parsed = Number(value.replace("$", "").replace("/hr", "").replace(",", "").trim());
    return Number.isFinite(parsed) ? parsed : fallback;
  }
  return fallback;
}

function asBoolean(value: unknown): boolean | null {
  return typeof value === "boolean" ? value : null;
}

function pickString(row: JsonRecord, keys: string[], fallback = ""): string {
  for (const key of keys) {
    const value = row[key];
    if (typeof value === "string" && value.trim()) {
      return value;
    }
  }
  return fallback;
}

function toSourceResult(name: string, state: DataState, message: string, url?: string): SourceResult {
  return { name, state, message, url };
}

async function fetchJson(url: string): Promise<unknown> {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), REQUEST_TIMEOUT_MS);
  try {
    const response = await fetch(url, {
      cache: "no-store",
      headers: { "User-Agent": "FishDashboard/0.2" },
      signal: controller.signal
    });
    if (!response.ok) {
      throw new Error(`HTTP ${response.status}`);
    }
    return response.json();
  } finally {
    clearTimeout(timeout);
  }
}

async function fetchPaged(url: string, listKey: "nodes" | "envs"): Promise<{ payloads: unknown[]; source: SourceResult }> {
  const payloads: unknown[] = [];
  let totalItems = 0;
  let lastMessage = "";
  for (let page = 1; page <= MAX_PAGES; page += 1) {
    const pageUrl = `${url}${url.includes("?") ? "&" : "?"}page=${page}`;
    const payload = await fetchJson(pageUrl);
    payloads.push(payload);
    const rows = isRecord(payload) ? asArray(payload[listKey]) : [];
    const pagination = isRecord(payload) && isRecord(payload.pagination) ? payload.pagination : {};
    totalItems = asNumber(pagination.totalItems, totalItems);
    const totalPages = asNumber(pagination.totalPages, page);
    lastMessage = `Fetched ${rows.length} ${listKey} on page ${page}${totalItems ? ` of ${totalItems}` : ""}`;
    if (rows.length === 0 || page >= totalPages) {
      break;
    }
  }

  return {
    payloads,
    source: toSourceResult("dashboard-api", "live", lastMessage || `Fetched ${payloads.length} pages`, url)
  };
}

function normalizeNodeCandidate(value: unknown): JsonRecord | null {
  if (!isRecord(value)) {
    return null;
  }
  if (isRecord(value._source)) {
    return value._source;
  }
  return value;
}

function nodeRowsFromPayload(payload: unknown): JsonRecord[] {
  if (!isRecord(payload)) {
    return [];
  }
  const rows = asArray(payload.nodes).length ? asArray(payload.nodes) : asArray(payload.envs);
  return rows.map(normalizeNodeCandidate).filter((row): row is JsonRecord => row !== null);
}

function nodeMetadataFromPayload(payload: unknown): NodeMetadata[] {
  return nodeRowsFromPayload(payload).map(nodeMetadataFromRow);
}

function nodeMetadataFromRow(node: JsonRecord): NodeMetadata {
  const providerId = pickString(node, ["id", "address", "friendlyName"], "unknown-provider");
  return {
    providerId,
    providerLabel: pickString(node, ["friendlyName", "address", "id"], providerId).slice(0, 42),
    region: regionForNode(node),
    eligible: asBoolean(node.eligible),
    eligibilityCause: asString(node.eligibilityCauseStr) || null,
    http: asBoolean(node.http),
    p2p: asBoolean(node.p2p),
    version: asString(node.version) || null,
    lastSeen: timestampToIso(node.lastSeen ?? node.timestamp),
    uptimeSeconds: Number.isFinite(asNumber(node.uptime, Number.NaN)) ? asNumber(node.uptime, 0) : null,
    nodeEndpoint: asString(asArray(node.currentAddrs)[0], "")
  };
}

function invalidGpuDescription(description: string): boolean {
  const value = description.toLowerCase();
  return (
    !value ||
    value.includes("failed to") ||
    value.includes("virtio") ||
    value.includes("cirrus") ||
    value.includes("blocked by the operating system")
  );
}

function isGpuResource(resource: JsonRecord): boolean {
  const type = asString(resource.type).toLowerCase();
  const description = pickString(resource, ["description", "name", "id"]);
  if (type !== "gpu") {
    return false;
  }
  if (invalidGpuDescription(description)) {
    return false;
  }
  const lower = description.toLowerCase();
  return /(nvidia|geforce|rtx|a100|h100|h200|l40|a6000|amd|radeon|mi300|gfx)/i.test(lower);
}

function priceForResource(environment: JsonRecord, resourceId: string): { value: number | null; feeToken: string; basis: ComputeResource["priceBasis"] } {
  const fees = isRecord(environment.fees) ? environment.fees : {};
  for (const feeGroup of Object.values(fees)) {
    for (const fee of asArray(feeGroup)) {
      if (!isRecord(fee)) {
        continue;
      }
      const feeToken = asString(fee.feeToken, "unknown");
      for (const price of asArray(fee.prices)) {
        if (!isRecord(price)) {
          continue;
        }
        if (asString(price.id) === resourceId) {
          const value = asNumber(price.price, Number.NaN);
          return {
            value: Number.isFinite(value) ? value : null,
            feeToken,
            basis: feeToken.toLowerCase() === BASE_USDC ? "configured_fee" : "unknown"
          };
        }
      }
    }
  }

  return { value: null, feeToken: "unknown", basis: "unknown" };
}

function regionForNode(node: JsonRecord): string {
  const location = isRecord(node.location) ? node.location : {};
  return pickString(location, ["region", "country", "city"], "unknown");
}

function timestampToIso(value: unknown): string | null {
  const timestamp = asNumber(value, Number.NaN);
  if (!Number.isFinite(timestamp) || timestamp <= 0) {
    return null;
  }
  const milliseconds = timestamp < 10_000_000_000 ? timestamp * 1000 : timestamp;
  return new Date(milliseconds).toISOString();
}

function normalizeNodeResources(node: JsonRecord, source: ComputeResource["source"], endpointUrl: string): ComputeResource[] {
  const providerId = pickString(node, ["id", "address", "friendlyName"], "unknown-provider");
  const providerLabel = pickString(node, ["friendlyName", "address", "id"], providerId).slice(0, 42);
  const region = regionForNode(node);
  const computeEnvironments = isRecord(node.computeEnvironments) ? node.computeEnvironments : {};
  const environments = asArray(computeEnvironments.environments).filter(isRecord);
  const timestamp = nowIso();
  const nodeEndpoint = endpointUrl || asString(asArray(node.currentAddrs)[0], "");

  return environments.flatMap((environment) => {
    const resources = asArray(environment.resources).filter(isRecord);
    return resources.filter(isGpuResource).map((resource) => {
      const resourceId = pickString(resource, ["id"], "gpu");
      const resourceName = pickString(resource, ["description", "name", "id"], "GPU");
      const total = asNumber(resource.total, asNumber(resource.max, 1));
      const inUse = asNumber(resource.inUse, 0);
      const available = Math.max(total - inUse, 0);
      const price = priceForResource(environment, resourceId);
      const environmentId = pickString(environment, ["id"], "unknown-env");

      return {
        snapshotId: `${timestamp}:${providerId}:${environmentId}:${resourceId}`,
        timestamp,
        source,
        providerId,
        providerLabel,
        nodeEndpoint,
        environmentId,
        region,
        resourceType: "gpu",
        resourceName,
        total,
        inUse,
        available,
        feeToken: price.feeToken,
        pricePerMinute: null,
        pricePerHour: null,
        listedPrice: price.value,
        priceBasis: price.basis,
        minJobDuration: asNumber(environment.minJobDuration, Number.NaN) || null,
        maxJobDuration: asNumber(environment.maxJobDuration, Number.NaN) || null,
        runningJobs: asNumber(environment.runningJobs, 0),
        status: available > 0 ? "available" : total > 0 ? "busy" : "unknown",
        raw: { node, environment, resource }
      } satisfies ComputeResource;
    });
  });
}

function dedupeResources(resources: ComputeResource[]): ComputeResource[] {
  const seen = new Map<string, ComputeResource>();
  for (const resource of resources) {
    const key = [
      resource.providerId,
      resource.environmentId,
      resource.resourceName,
      resource.total,
      resource.inUse,
      resource.listedPrice ?? "none"
    ].join("|");
    if (!seen.has(key)) {
      seen.set(key, resource);
    }
  }
  return Array.from(seen.values());
}

async function readNodeEndpoints(): Promise<string[]> {
  try {
    const raw = await readFile(NODE_ENDPOINTS_PATH, "utf8");
    return raw
      .split("\n")
      .map((line) => line.trim())
      .filter((line) => line && !line.startsWith("#"))
      .map((line) => line.replace(/\/$/, ""));
  } catch {
    return [];
  }
}

async function fetchDirectNode(endpoint: string): Promise<{ resources: ComputeResource[]; source: SourceResult }> {
  const url = `${endpoint}/api/services/computeEnvironments`;
  try {
    const payload = await fetchJson(url);
    const node = {
      id: endpoint,
      friendlyName: endpoint,
      currentAddrs: [endpoint],
      computeEnvironments: isRecord(payload) && Array.isArray(payload.environments) ? payload : { environments: payload }
    };
    const resources = normalizeNodeResources(node, "direct-node", endpoint);
    return {
      resources,
      source: toSourceResult("direct-node", "live", `Fetched ${resources.length} normalized GPU resources`, url)
    };
  } catch (error) {
    return {
      resources: [],
      source: toSourceResult("direct-node", "unavailable", error instanceof Error ? error.message : "Fetch failed", url)
    };
  }
}

async function loadSample(): Promise<OceanSummary> {
  const raw = await readFile(SAMPLE_PATH, "utf8");
  const parsed = JSON.parse(raw) as OceanSummary;
  return {
    ...parsed,
    dataState: "sample",
    lastUpdated: nowIso(),
    sourceCount: parsed.sources?.length ?? 1,
    warnings: [
      "Showing static sample data because live Oncompute/Ocean resources were unavailable.",
      ...(parsed.warnings ?? [])
    ]
  };
}

function median(values: number[]): number | null {
  if (!values.length) {
    return null;
  }
  const sorted = [...values].sort((a, b) => a - b);
  const middle = Math.floor(sorted.length / 2);
  if (sorted.length % 2 === 0) {
    return (sorted[middle - 1] + sorted[middle]) / 2;
  }
  return sorted[middle];
}

function summarize(resources: ComputeResource[], sources: SourceResult[], analytics: JsonRecord | null, nodeMetadata: Map<string, NodeMetadata>, providerPilot: ProviderPilotRegistry): OceanSummary {
  const grouped = new Map<string, ComputeResource[]>();
  for (const resource of resources) {
    const key = resource.resourceName;
    grouped.set(key, [...(grouped.get(key) ?? []), resource]);
  }

  const gpuSupply: GpuSupplyRow[] = Array.from(grouped.entries())
    .map(([gpu, rows]) => {
      const usdHrPrices = rows.map((row) => row.pricePerHour).filter((value): value is number => value !== null);
      const listedPrices = rows.map((row) => row.listedPrice).filter((value): value is number => value !== null);
      return {
        gpu,
        total: rows.reduce((sum, row) => sum + row.total, 0),
        available: rows.reduce((sum, row) => sum + row.available, 0),
        providers: new Set(rows.map((row) => row.providerId)).size,
        lowestUsdHr: usdHrPrices.length ? Math.min(...usdHrPrices) : null,
        medianUsdHr: median(usdHrPrices),
        lowestListedPrice: listedPrices.length ? Math.min(...listedPrices) : null,
        regions: Array.from(new Set(rows.map((row) => row.region).filter(Boolean))).sort()
      };
    })
    .sort((a, b) => b.available - a.available);

  const byProvider = new Map<string, ComputeResource[]>();
  for (const resource of resources) {
    byProvider.set(resource.providerId, [...(byProvider.get(resource.providerId) ?? []), resource]);
  }

  const providers: ProviderScore[] = Array.from(byProvider.entries())
    .map(([providerId, rows]) => {
      const listedPrices = rows.map((row) => row.listedPrice).filter((value): value is number => value !== null);
      const usdHrPrices = rows.map((row) => row.pricePerHour).filter((value): value is number => value !== null);
      const availableGpus = rows.reduce((sum, row) => sum + row.available, 0);
      const node = nodeMetadata.get(providerId);
      const pilotProfile = providerPilot.providers.find((provider) => provider.providerId === providerId);
      const allowlist = providerPilot.allowlist.find((entry) => entry.providerId === providerId);
      const selected = Boolean(allowlist || pilotProfile?.pilotStatus === "allowed");
      const nodeStatus: ProviderScore["nodeStatus"] = node?.eligible === true ? "eligible" : node?.eligible === false ? "not_eligible" : "unknown";
      const fishReadyStatus = fishReadyStatusForProvider({
        availableGpus,
        selected,
        readinessComplete: Boolean(pilotProfile && pilotProfile.readiness.readyCount === pilotProfile.readiness.totalCount)
      });
      return {
        providerId,
        label: node?.providerLabel ?? pilotProfile?.publicLabel ?? rows[0]?.providerLabel ?? providerId,
        region: node?.region ?? pilotProfile?.region ?? rows[0]?.region ?? "unknown",
        gpuTypes: Array.from(new Set(rows.map((row) => row.resourceName))).sort(),
        availableGpus,
        lowestUsdHr: usdHrPrices.length ? Math.min(...usdHrPrices) : null,
        lowestListedPrice: listedPrices.length ? Math.min(...listedPrices) : null,
        uptime7d: node?.uptimeSeconds ? Math.min(1, node.uptimeSeconds / (7 * 24 * 60 * 60)) : null,
        benchmarkStatus: selected ? "selected for benchmark" : "needs benchmark",
        pilotEligible: availableGpus > 0 && node?.eligible !== false,
        nodeStatus,
        fishReadyStatus,
        nodeHttp: node?.http ?? null,
        nodeP2p: node?.p2p ?? null,
        nodeVersion: node?.version ?? null,
        lastSeen: node?.lastSeen ?? null,
        readinessLabel: pilotProfile?.readiness.label ?? null,
        verified: selected
      };
    })
    .sort((a, b) => b.availableGpus - a.availableGpus);

  const h200Prices = resources
    .filter((row) => row.resourceName.toLowerCase().includes("h200"))
    .map((row) => row.pricePerHour)
    .filter((value): value is number => value !== null);
  const listedPrices = resources.map((row) => row.listedPrice).filter((value): value is number => value !== null);

  const totalNetworkJobs = asNumber(analytics?.totalNetworkJobs, 0);
  const totalBenchmarkJobs = asNumber(analytics?.totalBenchmarkJobs, 0);
  const totalNetworkRevenue = Number.isFinite(asNumber(analytics?.totalNetworkRevenue, Number.NaN))
    ? asNumber(analytics?.totalNetworkRevenue, 0)
    : null;

  return {
    dataState: "live",
    lastUpdated: nowIso(),
    sources,
    sourceCount: sources.filter((source) => source.state === "live").length,
    kpis: {
      totalGpus: resources.reduce((sum, row) => sum + row.total, 0),
      availableGpus: resources.reduce((sum, row) => sum + row.available, 0),
      providerCount: byProvider.size,
      eligibleNodeCount: providers.filter((provider) => provider.nodeStatus === "eligible").length,
      fishReadyProviderCount: providers.filter((provider) => provider.fishReadyStatus === "ready" || provider.fishReadyStatus === "selected").length,
      h200FromUsdHr: h200Prices.length ? Math.min(...h200Prices) : null,
      lowestListedGpuFee: listedPrices.length ? Math.min(...listedPrices) : null,
      oceanNativeJobs: totalNetworkJobs,
      benchmarkJobs: totalBenchmarkJobs,
      providerPayoutUsd: null,
      networkRevenueUsd: totalNetworkRevenue
    },
    gpuSupply,
    providers,
    warnings: [
      "Live Oncompute fee fields are shown as listed configured fees until price basis is confirmed.",
      "Virtual GPUs and known NVML error rows are filtered out of public supply metrics.",
      ...(providerPilot.allowlist.length ? [] : ["No Fish selected-provider allowlist is configured yet. Supply rows are candidates, not routed providers."])
    ]
  };
}

function fishReadyStatusForProvider(input: { availableGpus: number; selected: boolean; readinessComplete: boolean }): ProviderScore["fishReadyStatus"] {
  if (input.readinessComplete) {
    return "ready";
  }
  if (input.selected) {
    return "selected";
  }
  if (input.availableGpus > 0) {
    return "candidate";
  }
  return "needs_review";
}

export async function collectOceanData(): Promise<OceanData> {
  const sources: SourceResult[] = [];
  const resources: ComputeResource[] = [];
  const nodeMetadata = new Map<string, NodeMetadata>();
  let analytics: JsonRecord | null = null;
  const providerPilot = await collectProviderPilotRegistry();

  for (const [url, key] of [
    [process.env.ONCOMPUTE_NODES_URL ?? "https://api.oncompute.ai/nodes", "nodes"],
    [process.env.ONCOMPUTE_ENVS_URL ?? "https://api.oncompute.ai/envs", "envs"]
  ] as const) {
    try {
      const result = await fetchPaged(url, key);
      sources.push(result.source);
      for (const payload of result.payloads) {
        for (const node of nodeMetadataFromPayload(payload)) {
          nodeMetadata.set(node.providerId, node);
        }
        for (const node of nodeRowsFromPayload(payload)) {
          resources.push(...normalizeNodeResources(node, "dashboard-api", url));
        }
      }
    } catch (error) {
      sources.push(toSourceResult("dashboard-api", "unavailable", error instanceof Error ? error.message : "Fetch failed", url));
    }
  }

  const statsUrl = process.env.ONCOMPUTE_STATS_URL ?? "https://analytics.oncompute.ai/global-stats";
  try {
    const payload = await fetchJson(statsUrl);
    if (isRecord(payload)) {
      analytics = payload;
    }
    sources.push(toSourceResult("analytics-api", "live", "Fetched global network stats", statsUrl));
  } catch (error) {
    sources.push(toSourceResult("analytics-api", "unavailable", error instanceof Error ? error.message : "Fetch failed", statsUrl));
  }

  const directEndpoints = await readNodeEndpoints();
  for (const endpoint of directEndpoints) {
    const result = await fetchDirectNode(endpoint);
    sources.push(result.source);
    resources.push(...result.resources);
  }

  const normalized = dedupeResources(resources);
  if (normalized.length > 0) {
    return {
      summary: summarize(normalized, sources, analytics, nodeMetadata, providerPilot),
      resources: normalized
    };
  }

  const sample = await loadSample();
  return {
    summary: {
      ...sample,
      sources: [...sources, ...(sample.sources ?? [])],
      sourceCount: sources.filter((source) => source.state === "live").length
    },
    resources: []
  };
}
